lib/router/src/cache.rs

```diff
@@ -1,426 +1,397 @@
-        }
+            expires,
+            expirations: DelayQueue::with_capacity(capacity),
+            values: IndexMap::default(),
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
+
+        (cache, bg_purge)
     }
-}
 
-impl<K: Hash + Eq, V, N: Now> Cache<K, V, N> {
-    /// Accesses a route.
-    ///
-    /// A mutable reference to the route is wrapped in the returned `Access` to
-    /// ensure that the access-time is updated when the reference is released.
-    pub fn access<Q>(&mut self, key: &Q) -> Option<Access<V, N>>
-    where
-        Q: Hash + Equivalent<K>,
-    {
-        let v = self.vals.get_mut(key)?;
-        Some(v.access(&self.now))
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
 
-    /// Ensures that there is capacity to store an additional route.
-    ///
-    /// Returns a handle that may be used to store an ite,. If there is no available
-    /// capacity, idle entries may be evicted to create capacity.
-    ///
-    /// An error is returned if there is no available capacity.
-    pub fn reserve(&mut self) -> Result<Reserve<K, V, N>, CapacityExhausted> {
-        if self.vals.len() == self.capacity {
-            // Only whole seconds are used to determine whether a node should be retained.
-            // This is intended to prevent the need for repetitive reservations when
-            // entries are clustered in tight time ranges.
-            let max_age = self.max_idle_age.as_secs();
-            let now = self.now.now();
-            self.vals.retain(|_, n| {
-                let age = now - n.last_access();
-                age.as_secs() <= max_age
-            });
-
-            if self.vals.len() == self.capacity {
-                return Err(CapacityExhausted {
-                    capacity: self.capacity,
-                });
-            }
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        if let Some(node) = self.values.get_mut(key) {
+            self.expirations.reset(node.dq_key_ref(), self.expires);
+            return Some(node.value_ref().clone());
         }
 
-        Ok(Reserve {
-            vals: &mut self.vals,
-            now: &self.now,
-        })
+        None
     }
 
-    /// Overrides the time source for tests.
-    #[cfg(test)]
-    fn with_clock<M: Now>(self, now: M) -> Cache<K, V, M> {
-        Cache {
-            now,
-            vals: self.vals,
-            capacity: self.capacity,
-            max_idle_age: self.max_idle_age,
-        }
+    pub fn insert(&mut self, key: K, value: V) -> Option<Node<V>> {
+        let node = {
+            let dq_key = self.expirations.insert(key.clone(), self.expires);
+            Node::new(dq_key, value)
+        };
+
+        self.values.insert(key, node)
     }
-}
 
-// ===== impl Reserve =====
+    pub fn poll_insert(&mut self) -> Poll<(), ()> {
```

Comment:
I don't think it makes a lot of sense for this to return a `Poll`. Looking at how it's used, it looks like we `expect` that it will always be `Ok`, and return a `NoCapacity` error if it's `NotReady`: 
https://github.com/linkerd/linkerd2-proxy/blob/39f00f9753b8c1affb070cf1a8cc8ef24b2ae3af/lib/router/src/lib.rs#L283-L295

This seems unnecessarily complex to me. In particular, I notice that this method will never return `Err` --- we panic if polling the `DelayQueue` failed, and there's no other case where we would return `Err`, so wrapping this in a `Result` that we expect will never be `Err` is unnecessary. In the router code where we call this method, rather than yielding when we see a `NotReady`, we return an error, so perhaps _that_ case ought to be represented using `Err`.

I think we should consider changing this from `poll_insert` to something like
```rust
pub fn can_insert(&mut self) -> Result<(), NoCapacity>
```
If this had that signature, the case in `ResponseFuture::poll` could just look like this:
```rust
if let Some(service) = cache.get(&target) {
    State::Calling(Some(request), Some(service))
} else {
    // Ensure that there is capacity for a new slot
    cache.can_insert()?;
    // Make a new service for the target
    let make = rr.make.take().expect("polled after ready");
    let service = LoadShed::new(make.make(&target));

    cache.insert(target, service.clone());
    State::Calling(Some(request), Some(service))
}
```

Reply:
I'll look into your suggestion and see how it changes things, but before doing so I'll leave a quick explanation on why I ultimately thought `Poll` was a good choice.

In a previous state of this PR, `poll_insert()` (basically what `reserve()` was) did return a `Result`. However Sean pointed out [here](https://github.com/linkerd/linkerd2-proxy/pull/247#discussion_r284857061) that it internally calls `poll()` and could therefore panic if called outside a task. I felt like that was an important point, and it still holds true even though a lot else has changed.

I know the cases of this being used elsewhere where the implementer doesn't know to only call it within a task, but maybe something to keep in mind.

Reply:
> However Sean pointed out [here](https://github.com/linkerd/linkerd2-proxy/pull/247#discussion_r284857061) that it internally calls `poll()` and could therefore panic if called outside a task. I felt like that was an important point, and it still holds true even though a lot else has changed.

That's a good point from @seanmonstar that I hadn't thought of. I think there's some tension here between indicating that the function should not be called outside of a task context, and simplifying the signature of this function. It definitely seems strange to have a function called `poll_XXX` that returns a `Result`, so the return type of `Poll` makes sense in that context, but it also seems to introduce some complexity downstream that's otherwise unnecessary.

I think either way, there should be a doc comment stating that this shouldn't be called outside a task context, whether it remains `poll_insert` or not.

Reply:
> I think what we should probably do is leave the function poll_insert but change it to return Poll<(), NoCapacity>. 

Yep I like that approach. It will be nice to use the `?` operator downstream, but keep a signature that indicates this should be called within a task.